implementations/rust/ockam/ockam_core/src/access_control.rs

```diff
@@ -32,66 +32,18 @@ pub trait AccessControl: Debug + Send + Sync + 'static {
     async fn is_authorized(&self, relay_msg: &RelayMessage) -> Result<bool>;
 }
 
```

Comment:
What is the benefit of that?

Reply:
It gives client code flexibility. Assume you have a value of a type that implements `AccessControl` in an `Arc` and you want to call some API that accepts any implementation of `AccessControl`. Without a forwarding impl you can not call the function with the value you have as `Arc<T>` for `T: AccessControl` is not itself an implementation of `AccessControl`. Similarly for `Box` or shared references.

Reply:
Tnx for the explanation. I decided to use dynamic dispatch for `AccessControl`, to support changing type of the implementation in the runtime, e.g. a `Context` could want to change its access control from `DenyAll` to `AllowAll`. Because of that I think we should pass Access Control only as a reference or wrapped in an `Arc`. And I think it's advisable to be able to distinguish between `T` and `Arc<T>`, for example in cases like your suggestion above: to let `ctx.start_worker` accept `impl AccessControl`. If `Arc<T>` is passed their, we'll end up creating `Arc<Arc<T>>`, don't we?

Reply:
`Arc<Arc<T>>` is not wrong. Another question is of course why the `Arc` is needed at all? Maybe a `Clone` constraint would be better to express the need to create duplicates?

Reply:
How would you use `Clone` in dynamic dispatch usecase?

Reply:
I only glanced over the bit about storing trait objects, but yes, for storing a value as a field without involving generics some indirection is needed.

Reply:
That's why I decided to go with `Arc` for now

Reply:
I still do not understand why you want to prevent someone with an `Arc<T>` or `Box<T>` where `T` implements `AccessControl` to pass that to functions that are generic over `AccessControl` types, but :shrug: